### ISSUE TITLE
Gatekeeper requires the `iss`-claim in the auth-token

### ISSUE THREAD

**Issue by n0emis** [gogatekeeper/gatekeeper#11]
# Gatekeeper requires the `iss`-claim in the auth-token

## Summary 
Gatekeeper does not work with SimpleSamlPHP (with the OIDC-Module enabled) as a IdP. 

Gatekeeper requires the `access_token` to have specific information (for example groups, the `iss`-claim etc..). 

Following the OpenID Connect Spec, all this data should be in the `id_token` (https://openid.net/specs/openid-connect-core-1_0.html#IDToken). But since gatekeeper only uses the `access_token`, this data will not be used. Therefor gatekeeper is not fully spec-compliant.

## Environment

- OS: Linux
- Kernel (e.g. `uname -a`): 5.9.13-arch1-1
- Go (e.g. go version): go1.15.6
- Server (e.g. Keycloak or any other IdP): SimpleSamlPHP
- Louketo: 1.0.0

## Expected Results
Logging in should work

## Actual Results
```
error	louketo-proxy/handlers.go:152	unable to verify the id token	{"error": "oidc: JWT claims invalid: missing claim: 'iss'"}
info	louketo-proxy/middleware.go:146	client request	{"latency": 0.881592596, "status": 403, "bytes": 0, "client_ip": "127.0.0.1:60538", "method": "GET", "path": "/oauth/callback"}
```

## Steps to reproduce

1. Configure gatekeeper like this:
```yaml
client-id: _9e76c54497415dfbd2774738c5ca8ba4abd7ddfec6
discovery-url: https://simplesaml.n0emis.tk
encryption_key: foobar2342
secure-cookie: false
listen: :3000
redirection-url: http://127.0.0.1:3000
upstream-url: http://127.0.0.1:8000

resources:
- uri: /*
```

2. Try to log in

## Additional Information

I will leave simplesamlphp configured and this client enabled so that other people can test. (It requires no authentification, since it is configured to only returns static userinfo)

**Comment by n0emis** [comment]
When commenting this part out, it suddenly works. Although then roles from keycloak are no longer detected.

https://github.com/go-gatekeeper/gatekeeper/blob/b043dea3a21d9ca5955edbe194b572db9b7d6b8c/handlers.go#L142-L148

Maybe some kind of config option could solve this?

**Comment by p53** [comment]
yup, there was already issue with old louketo for this, access token is verified instead of id token, per spec

**Comment by p53** [comment]
problem is more complex, gatekeeper fullfils two things 1. is authentication -> authorization code flow 2. authorization (authorizing access tokens), with 1. you need according spec to validate only id token but we are also doing 2. and in that case we need to verify also access tokens - question is how, currently in code, you commented, gatekeeper uses oidc library, which is intended for verification of id tokens, to verify access token so because of that we see errors like yours and others e.g. expecting in audience same principal as in client_id....

So for me solution should include, but requires discussion and probably will be more complex:
1. verify id_token always if present - case of authorization code flow
2. verify access_token optionally in case of authorisation code flow
3. verify signature always in case access_token as bearer token
4. verify other fields optionally in case of bearer token
5. question is if don't offload authorization to some external service through plugin

@fredbi @ackerleytng @statik

**Comment by ackerleytng** [comment]
Perhaps upgrading to the current version of https://github.com/coreos/go-oidc would change things here, and allow gatekeeper to be more spec-compliant. The `IdentityFromClaims` function seems to have been removed from the current version.

One problem here is that current users of gatekeeper might be relying on this check for authorization, so changing to be more spec-compliant actually kind of reduces security for existing users who are using this for authorization.

**Comment by p53** [comment]
@ackerleytng yup oidc is all over the code (i already tried to update it) we will need also to update jose library...it will be quite big change->it would be good to have integration tests before such changes (quite a lot of work), but would still solve this problem only partially, we would need to introduce some switches for on/off authorization (by default can be on to be backward compatible)

**Comment by damien-d-gerard** [comment]
Hi guys,
Is there any plan to upgrade on go-oidc v2?
I'm still blocked on this issue https://github.com/louketo/louketo-proxy/issues/567 without a solution 😭 

@p53 maybe the solution is to create a new Major Branch with no backward compatiblility ?

**Comment by p53** [comment]
@damien-d-gerard i had same issue with keycloak, you can workaround this by adding audience to your client - at least in keycloak, this issue doesn't require to upgrade to go-oidc v2 problem is that code verifies access_token instead of id token with method designed for id token, we should be able relatively easily to fix this but this will create incompatibility+not verifying access token which is thing many people might need and expect. Right now there is no plan, but it will be one of high priority issues (already in issues) in next releases together with other libraries

**Comment by p53** [comment]
@damien-d-gerard @n0emis could you check this pull request https://github.com/gogatekeeper/gatekeeper/pull/20, try to build and test on your use case

**Comment by n0emis** [comment]
No, unfortunately not: 
```
error	access token failed verification	{"client_ip": "127.0.0.1:53916", "error": "oidc: id token issued by a different provider, expected \"https://simplesaml.n0emis.tk\" got \"\""}
```

**Comment by p53** [comment]
@n0emis sorry forgot to turn off issuer check, now is there https://github.com/gogatekeeper/gatekeeper/pull/20/commits/498361d14f363d36a1e410cbac4ce2b0392e8257

**Comment by n0emis** [comment]
Thanks, now it works

**Comment by p53** [comment]
@n0emis @damien-d-gerard @akifkhan01 - there is 1.2.0-rc1 release please try out, there are two new flags SkipAccessTokenIssuerCheck, SkipAccessTokenClientIDCheck

@vxavictor513 @Adriien-M - i think this issue https://github.com/louketo/louketo-proxy/issues/670 should be resolved in our 1.2.0-rc1 release
